subsys/net/lib/sockets/sockets.c

```diff
@@ -332,17 +345,21 @@ static inline ssize_t zsock_recv_dgram(struct net_context *ctx,
 		}
 	}
 
-	/* Remove packet header since we've handled src addr and port */
+	/* Set starting point behind packet header since we've
+	 * handled src addr and port.
+	 */
 	header_len = net_pkt_appdata(pkt) - pkt->frags->data;
-	net_buf_pull(pkt->frags, header_len);
```

Comment:
Why do we remove the net_buf_pull() here, isn't the IP header left now there as is?

Reply:
When using MSG_PEEK we must not pull the header because it may be fetched in a later recv. I changed the code to use header_len as offset in net_frag_linearize() instead.  This way we don't need different code for peek and non-peek.